src/effects/deeplearning/ModelCard.cpp

```diff
@@ -0,0 +1,272 @@
+   {
+      std::stringstream msg;
+      msg << "field: " << key << " is not of type: " << type;
+      throw InvalidModelCardDocument(msg.str(), doc);
+   }
+
+   std::string tryGetString(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsString())
+         throwTypeError(key, "string", doc);
+
+      return (*doc)[key.c_str()].GetString();
+   }
+
+   int tryGetInt(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsInt())
+         throwTypeError(key, "int", doc);
+      
+      return (*doc)[key.c_str()].GetInt();
+   }
+
+   bool tryGetBool(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsBool())
+         throwTypeError(key, "bool", doc);
+      
+      return (*doc)[key.c_str()].GetBool();
+   }
+
+   std::vector<std::string> tryGetStringArray(const std::string &key, DocHolder doc)
+   {
+      validateExists(key, doc);
+      if(!(*doc)[key.c_str()].IsArray())
+         throwTypeError(key, "array", doc);
+      
+      std::vector<std::string> labels;
+      for (auto itr = (*doc)[key.c_str()].Begin(); itr != (*doc)[key.c_str()].End();  ++itr)
+         labels.emplace_back(itr->GetString());
+
+      return labels;
+   }
+
+}
+
+namespace parsers 
+{
+   DocHolder ParseString(const std::string &data)
+   {
+      DocHolder d = std::make_shared<rapidjson::Document>();
+      // parse the data
+      d->Parse(data.c_str());
+      if (d->Parse(data.c_str()).HasParseError()) 
+      {
+         std::string msg = "error parsing JSON from string: \n";
+         msg += rapidjson::GetParseError_En(d->GetParseError());
+         msg += "\n document: \n\t" + std::string(data.c_str()) + "\n";
+         throw InvalidModelCardDocument(msg.c_str(), d);
+      }
+      
+      return d;
+   }
+
+   DocHolder ParseFile(const std::string &path)
+   {
+      // TODO: error check reading from the stream?
+      // TODO: error check for file parse
+      // https://rapidjson.org/md_doc_stream.html
+      FILE* fp = fopen(path.c_str(), "r");
+
+      char readBuffer[65536];
+      rapidjson::FileReadStream is(fp, readBuffer, sizeof(readBuffer));
+
+      DocHolder d;
+      d->ParseStream(is);
+
+      fclose(fp);
+
+      return d;
+   }
+}
+
+// ModelCard Implementation
+
+// copy constructor
+// copy all of the members of ModelCard
+ModelCard::ModelCard(const ModelCard &other)
+: m_name(other.name()),
+  m_author(other.author()),
+  m_description(other.description()),
+  m_version(other.version()),
+  m_sample_rate(other.sample_rate()),
+  m_multichannel(other.multichannel()),
+  m_effect_type(other.effect_type()),
+  m_domain(other.domain()),
+  m_labels(other.labels())
+{
+   
+}
+
+ModelCard::ModelCard()
+{
+
+}
+
+void ModelCard::SerializeToFile(const std::string &path) const
+{
+   std::ofstream ofs(path);
```

Comment:
Similar as above. You can implement your own writer or even simply serialize to string first